master/internal/api_trials.go

```diff
@@ -546,17 +550,12 @@ func (a *apiServer) TrialPreemptionSignal(
 	}
 
 	id := uuid.New()
-	ret, err := a.askAtDefaultSystem(trial, trialWatchPreemption{id: id})
-	if err != nil {
+	var signal <-chan bool
```

Comment:
A mixed use of go channel and actor actually drives me crazy. We can create an actor for each `TrialPreemptionSignalRequest` to listen to the trial actor and make the trial actor notify all of the listening actors of the preemption signals.

Reply:
You don't have to change it now. I just thought there was a more understandable way to write this but you might disagree.

Reply:
ah, i had a really long discussion with @rb-determined-ai about this. honestly, channels or actors 🤷 it really doesn't make a huge difference, but then you have an extra actor vs an extra channel for each caller, and you have to write the little actor. after writing a bit more go concurrency code, i think this is probably more approachable.

Reply:
I took the actor approach initially with trial log streaming and backed out of it because it added so much unnecessary complexity compared to channels.

Reply:
The channel code is more "go native". If we hire more go systems engineers, this will be more approachable right off, I bet.

Reply:
> but then you have an extra actor vs an extra channel for each caller, and you have to write the little actor. after writing a bit more go concurrency code, i think this is probably more approachable.

The idea of using actor is to avoid adding synchronization (channel blocking) so that one day it wouldn't get out of control. I think adding one actor wouldn't add more effort in this case. It just needs more thinking. We used to use channels everywhere and moved to an actor system for a reason.

> I took the actor approach initially with trial log streaming and backed out of it because it added so much unnecessary complexity compared to channels.

IDK the log streaming but this is definitely not the case here. Adding one actor type alone doesn't add complexity compared to the complexity added by one synchronization point. It's just hard to understand unless you look at the code in all places that involve this synchronization.

> The channel code is more "go native". If we hire more go systems engineers, this will be more approachable right off, I bet.

This is definitely not true. I guess the "go native" way isn't really appropriate in our applications. The mixed use of "go native" channels and actors just makes things more confusing.

Reply:
@stoksc can we at least use typed channels here so that we could use search to help understand the codebase?

Reply:
> The idea of using actor is to avoid adding synchronization (channel blocking) so that one day as you add more channels it wouldn't get out of control.

There is no synchronization here, we use a buffered channel that is the size of the max number of channels. The flipside of the actor system, and one we luckily don't hit very often, is that it's essentially like using `make(chan interface{}, ∞)` so we have no backpressure mechanism. If you wanted to, a send in a for loop to a slow actor would likely OOM the system.

> The other benefit is that since actor messages are strictly typed you can just search the message type to know how different parts of the actors work together.

This totally isn't true and is one of my biggest pet peeves about the actor system - there are multiple actors that receive a `model.State` from multiple places. It is impossible to statically reason about where those messages come from the way you describe. At least with my code, I send a specifically typed actor message (trialWatchPreemptionSignal that responds with a chan bool. It's fairly easily to trace where that chan comes from and what is in it.

> I think adding one actor wouldn't add more effort in this case. It just needs more thinking. We used to use channels everywhere and moved to an actor system for a reason.

I didn't know this, but I think for a case as simple as this, this is simpler personally. For long running processes that need to send and receive messages in an addressable way, actors totally make sense, but this is just propagating a bool 2 times.

> IDK the log streaming but this is definitely not the case here. Adding one actor type alone doesn't add complexity compared to the complexity added by one synchronization point. It's just hard to understand unless you look at the code in all places that involve this synchronization.

like I said, there is no synchronization point to do with channels like you describe. the handler asks the actor for a chan to receive signals on (the _ask_ is a synchronization point between the handler and the actor) and then the trial sends and handler receives at whim on the channel without any possible blocking.

> can we at least use typed channels here so that we could search to help understand the codebase?

what do you mean typed channels? a channel that sends a `type PreemptSignal bool` or something instead of bool?

Reply:
> We used to use channels everywhere and moved to an actor system for a reason.

Out of curiosity I git blamed pedl to see and found no evidence of this. It looked like we used globals accessed by API endpoints under a mutex and moved straight to an actor system.

Reply:
Some small bits too:
* it's much nicer IMO too to `select` off a grpc `ctx.Done()` and whatever is chan feeding data to your API endpoint (and whatever other signals), than to need to check `ctx.Err()` or make a go routine that selects off `ctx.Done()` and tells your intermediary actor.
* it is good to have to explicitly handle backpressure (channels have bounded cap), than "actor magic" where someone could easily oom the master filling up an actor's inbox.